DVCodingStyle.md

```diff
@@ -454,13 +455,194 @@ utility macros for even more common code sequences.
+        automatic function void get_nco();
+          ...
+        endfunction
+        // Bad: Putting functions and tasks in a header file that is potentially
+        //      included multiple times.
+
+        `define DV_FOO(...) ...
+        // Bad: No ifndef guard. Macro re-definition warnings will be thrown if
+        //      this header file is included in multiple sources.
+        ```
+
+        :+1:
+        ```systemverilog
+        // file: dv_macros.svh
+        `ifndef DV_FOO
+          `define DV_FOO(...) ...
+        `endif
+
+        `ifndef DV_BAR
+          `define DV_BAR xyz
+        `endif
+        ```
+
+        Generally, there is no need to `` `include`` this header file if it
+        can be guaranteed that it is already compiled prior to the current
+        compilation unit in which those macros are invoked. This is typically
+        handled by an external tool that works out the dependencies across
+        the compilation units (example: FuseSoC). However, it may not always be
+        possible to guarantee this, and the `` `include`` may be necessary.
+        The `` `ifndef`` guard helps prevent the re-definition of the macros with
+        itself.
+
+
+    *   If the scope of the macros defined is limited to a package or source
+        (individual interface, module or a class), then place them at the top of
+        the file. Always `` `undef`` them at the end of the file so that they
+        are no longer visible to subsequent sources during compilation. Do not
+        put `` `ifndef`` guards on these macros, so that the re-definitions
+        if encountered, are flagged.
+
+        :-1:
+        ```systemverilog
+        // file: foo_init_seq.sv
+        `ifndef GET_DATA
+          `define GET_DATA(baz) ...
+        `endif
+        // Bad: Macro-redefinition if encountered, will be ignored due to the
+        //      ifndef guard. Previously defined macro is invoked instead.
+        // Bad: Macro name is not specific enough.
+
+        class foo_init_seq;
+          ...
+        endclass
+        // EOF
+        // Bad: Macro is not undefined at the end of the file.
+        ```
+
+        :+1:
+        ```systemverilog
+        // file: foo_init_seq.sv
+        `define GET_FOO_TX_FIFO_DATA_AFTER_OP(baz) ...
+
+        class foo_init_seq;
+          ...
+        endclass
+
+        `undef GET_FOO_TX_FIFO_DATA_AFTER_OP
+        ```
+
+        :+1:
+        ```systemverilog
+        // file: foo_env_pkg.sv
+        `define GET_FOO_TX_FIFO_DATA(baz) ...
+
+        ...
+
+        // packages sources
+        `include "foo_env_cfg.sv"
+        ...
+        `include "foo_scoreboard.sv"
+        `include "foo_env.sv"
+
+
+        `undef GET_FOO_TX_FIFO_DATA
+        // Good: All package sources can invoke this macro. It will no longer
+        //       be visible beyond this compilation unit.
+        ```
+
+3.  Wrap macros that resolve to multiple statements within `begin` and `end`.
```

Comment:
The paper you cite actually suggests a `do begin ... end while (0)` loop, which also avoids the ugly problem of needing semicolons after some macros, but not others. Is there a reason not to include that?

Reply:
Wrapping the macro defs inside `do begin` ... `end while (0)` will require you to put a semi-colon at the end of a block of code. The following are the reasons for not mandating it:

- A block of code does not need semi-colon at the end. SV allows us to put arbitrary `begin..end` sections unlike C where arbitrary `{...}` would be a syntax error. 
- It makes it easy to distinguish macros that expand to a block of code (no semi-colon) vs macros that expand to an expression (semi-colon added). The ugliness you mention actually serves as an identification.
- This is in-line with UVM, which also does not require us to put semi-colon at the end of macros that expand into a block of code.
- Our OT codebase uses this style all over the place, so if we mandate `do begin ... end while (0)` it will be hard to update our codebase to match this style.  
- If we explicitly say  `do begin ... end while (0)` is also allowed, the 2nd bullet will no longer be true, and that will make the problem even uglier.

Reply:
Hmm, ok. I think I probably prefer semicolons everywhere but I guess UVM has made that decision for us already.